components/waypoint-movement.js

```diff
@@ -0,0 +1,239 @@
+      this.positions = new Array(this.children.length);
+      for (var i = 0; i < this.children.length; i++) {
+        this.positions[i] = new Float32Array(3);
+        this.children[i].getTranslationWorld(this.positions[i]);
+      }
+      this.currentPositionIndex = 0;
+      this.currentCurveIndex = 0;
+      this.currentLength = 0;
+      this._setCurvePoints();
+
+      this.fromPosition = this.positions[0];
+      this.toPosition = this.positions[1];
+      glMatrix.vec3.sub(this.currentPosition, this.toPosition, this.fromPosition);
+      this.fromTolength = glMatrix.vec3.length(this.currentPosition);
+    },
+    update: function(dt) {
+      this.currentLength += dt * this.speed;
+      let factor = this.currentLength / this.fromTolength;
+      if(factor > 0.5 && this.currentPositionIndex != this.positions.length - 1) {
+        this.incrementCurveIndex = true;
+      }
+      if(Math.abs(factor - Math.round(factor)) < this.curveDistance && !(this.currentPositionIndex == this.children.length-2 && factor > 0.5) && !(this.currentPositionIndex == 0 && factor < 0.5)) {
+        let bezFactor = (factor - Math.round(factor) + this.curveDistance) * this.bezFactorMultiplicator;
+        glMatrix.vec3.bezier(this.currentPosition, this.p0, this.p1,this.p1, this.p2, bezFactor);
+        glMatrix.vec3.bezier(this.lookPosition, this.p0, this.p1,this.p1, this.p2, bezFactor+0.01);
+      } else {
+        if(this.incrementCurveIndex && factor < 0.5 && factor > this.curveDistance && this.currentCurveIndex != this.positions.length - 3) {
+          this.currentCurveIndex++;
+          this.incrementCurveIndex = false;
+          this._setCurvePoints();
+        }
+        glMatrix.vec3.lerp(this.currentPosition, this.fromPosition, this.toPosition, factor);
+        glMatrix.vec3.lerp(this.lookPosition, this.fromPosition, this.toPosition, factor+0.01);
+      }
+      this.object.resetTranslationRotation();
+      lookAt(this.quat, this.lookPosition, this.currentPosition, this.up);
+      this.object.rotate(this.quat);
+      this.object.setTranslationWorld(this.currentPosition);
+      if(this.currentLength > this.fromTolength){
+        this.start = false;
+        this.currentPositionIndex++
+        if(this.currentPositionIndex == this.children.length-1) {
+          this.currentPositionIndex = 0;
+          this.currentCurveIndex = 0;
+          this._setCurvePoints();
+          this.incrementCurveIndex = false;
+          this.onFinalWaypointReachedCallbacks.forEach(function(f){ f() });
+        }
+        this.fromPosition = this.positions[this.currentPositionIndex];
+        this.toPosition = this.positions[this.currentPositionIndex + 1];
+        glMatrix.vec3.sub(this.currentPosition, this.toPosition, this.fromPosition);
+        this.fromTolength = glMatrix.vec3.length(this.currentPosition);
+        this.currentLength = 0;
+      }
+    },
+});
+
+let forwardTemp = new Float32Array(3);
```

Comment:
To avoid clashes with other scripts, you could wrap this into a function scope like so:

```js
const lookAt = function() {
    let temp = new Float32Array(3);
    // ...
    
    return function(quaternion, sourcePoint, ...) {
        // ... unchanged
    };
};
```

That way all the temp memory is private to the lookAt function and the variable names are not forever reserved in global namespace (due to let).

Also, I believe you wanted to add a comment that notes that this function will very soon be added to Wonderland Engine's JS API and therefore can soon be replaced 🎉

Reply:
The comment is added as a note in the readme, but I can also add it

Reply:
Will declaring the variables in the functions body not lead to issues with garbage collection?

Reply:
I fixed the code snippet. The idea is that the encapsulating function's scope has the one-time initialized temp memory and therefore you have the same effect for performance but without polluting the global scope.

Reply:
Perfect, thank you!